### PR TITLE
refactor(net): removes direct ui invocation from filemgr thread

### LINKED ISSUE #20428: Remove use of tinyfiledialogs from network code
Right now the engine is responsible for making UI appear to allow the user to choose an input file: https://github.com/servo/servo/blob/34f388229df4f201302ee51a8969c0556692bf0b/components/net/filemanager_thread.rs#L39-L67

Instead, the network code should send a message to the constellation, which can talk to the embedder and receive the final selection to return to the network code. This means the FileManager needs to have access to a channel which can communicate with the constellation; this can come from [CoreResourceManager::new](https://github.com/servo/servo/blob/34f388229df4f201302ee51a8969c0556692bf0b/components/net/resource_thread.rs#L303)  via [new_resource_threads](new_resource_threads). [This](https://github.com/servo/servo/blob/f467bdce1ba95e950b01f59ba284873137bca5d5/components/constellation/constellation.rs#L1845-L1851) is a good model for communicating with the embedder from the constellation.

To test this code, verify that it's possible to submit an HTML form that includes a `<input type="select">` element after choosing a file.

### PR DESCRIPTION
**Pull request by kwonoj**
- relates to https://github.com/servo/servo/issues/20428.

This PR tries to refactor `net::filemanager_thread` implementation, removes direct ui-related code invocation but ask constellation. I believe overall organization might need to be refactored still as I took my own liberty to wire up dots and dots between components, which could be non-recommended practices.

Probably point of review / need to be updated are

1. Communication between components 
Currently it's wired as like below:
```
+----------------+                    +---------------+              +---------+
|                |  constellationMsg  |               | embedderMsg  |         |
| filemgr_thread +------------------->+ constellation +------------->+ browser |
|                |                    |               |              |         |
+-----+----------+                    +---------------+              +----+----+
      ^                                                                   |
      +-------------------------------------------------------------------+
                                filelist: Vec(String)
```
- is this feasible approach?
- does organization of message / fn (where to put consteallation / embedder msg & fns) are legit?

2. Removal of `filemanger_thread::UIProvider`
- As filemanager_thread no longer need to aware actual ui, this PR removes `UIProvider` completely and let invoke tinyfiledialog directly in message listener. Maybe UIProvider itself still being needed?

3. Overall fn organization
- To reduce duplicated code it takes single msg with boolean flag to distinguish selecting multiple files, may feasible / or better to create explicit paths between two.

4. Invoking tfd in a separate thread
- This was mainly to align behavior to previous implentation, where tfd was invoked inside of filemanager_thread so does not block main. It may possibly just let block instead.

and of course, a lot of other codes need to follow better practices.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #20428 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____
- Manually verified file picker `<input type=files>` can pick up files.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20480)

### CODE REVIEW

**Comment by paulrouget** [comment]
I see you call `open_file_dialog` from a different thread to not block the main thread. Does that work well on the 3 platforms? (I can try on Windows and Mac if that helps).

**Comment by kwonoj** [comment]
@paulrouget I just give a quick try (open tfd for single / multiple dialog, check if it unblocks main) on linux / windows / macOS, and all platform seems working as expected. (Does not block / no immediate panic or unexpected behavior).

File: components/compositing/compositor_thread.rs [hunk 1]
```diff
@@ -141,6 +142,8 @@ pub enum EmbedderMsg {
     LoadComplete(TopLevelBrowsingContextId),
     /// A pipeline panicked. First string is the reason, second one is the backtrace.
     Panic(TopLevelBrowsingContextId, String, Option<String>),
+    /// Open file dialog to select files
```
Comment (paulrouget) [inline components/compositing/compositor_thread.rs]: Can you add a comment about what that boolean is for?

File: components/net/filemanager_thread.rs [hunk 2]
```diff
@@ -279,21 +219,39 @@ impl FileManagerStore {
...
 
-    fn select_file<UI>(&self,
-                       patterns: Vec<FilterPattern>,
-                       sender: IpcSender<FileManagerResult<SelectedFile>>,
-                       origin: FileOrigin,
-                       opt_test_path: Option<String>,
-                       ui: &UI)
-        where UI: UIProvider,
-    {
+    #[cfg(any(target_os = "macos", target_os = "linux", target_os = "windows"))]
```
Comment (paulrouget) [inline components/net/filemanager_thread.rs]: I don't think we have to make a distinction between the platforms here. Let send a message to the embedder no matter the platform, and let the embedder decide what to do depending on the platform.

File: components/constellation/constellation.rs [hunk 3]
```diff
@@ -1076,6 +1077,9 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
             FromCompositorMsg::SetCursor(cursor) => {
                 self.handle_set_cursor_msg(cursor)
             }
+            FromCompositorMsg::OpenFileSelectDialog(patterns, multiple_files, sender) => {
```
Comment (paulrouget) [inline components/constellation/constellation.rs]: Having to use `FromCompositorMsg` is unfortunate. Maybe a `Request::Script` might be better? Or even its own type of request.

**Review by paulrouget — CHANGES_REQUESTED** [review]
Could we short-circuit the constellation and directly use the embedder channel (EmbedderProxy)?

**Comment by kwonoj** [comment]
> Could we short-circuit the constellation and directly use the embedder channel (EmbedderProxy)?

@paulrouget I haven't tried but it may possible, mostly I tried follow steps in original issue https://github.com/servo/servo/issues/20428 's `network code should send a message to the constellation, which can talk to the embedder and receive the final selection to return to the network code` to let constellation handles it. May I update PR as suggested?

**Comment by paulrouget** [comment]
So far, only the constellation talks to the embedder via EmbedderMsg. So maybe let's keep it the way you did.

**Comment by kwonoj** [comment]
@paulrouget appreciate for feedback, I tried to update PR as suggested

- create FileManagerMsg and let it handled via Script::Request
- move out platform-aware logic from filethread

If there are things need to be corrected, I'll update per suggestions.

**Comment by paulrouget** [comment]
Sorry. Was away for a while. I'll look at this again soon.

**Comment by kwonoj** [comment]
no worries. 👍

**Comment by paulrouget** [comment]
I had a quick chat with @cbrewster and we agree that it's fine if the script thread has access to the EmbedderProxy. We could short-circuit the constellation.

File: components/compositing/compositor_thread.rs [hunk 4]
```diff
@@ -141,6 +142,8 @@ pub enum EmbedderMsg {
     LoadComplete(TopLevelBrowsingContextId),
     /// A pipeline panicked. First string is the reason, second one is the backtrace.
     Panic(TopLevelBrowsingContextId, String, Option<String>),
+    /// Open file dialog to select files. Set boolean flag to true allows to select multiple files.
+    GetSelectedFiles(Vec<FilterPattern>, bool, IpcSender<Option<Vec<String>>>),
```
Comment (paulrouget) [inline components/compositing/compositor_thread.rs]: Maybe rename that to `PickFiles`, `RequestFiles` or `SelectFiles`?

File: components/constellation/constellation.rs [hunk 5]
```diff
@@ -914,6 +931,15 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
...
     }
 
+    fn handle_request_from_filemanager(&mut self, message: FileManagerMsg) {
+        match message {
+            FileManagerMsg::OpenFileSelectDialog(patterns, multiple_files, sender) => {
+                let msg = EmbedderMsg::GetSelectedFiles(patterns, multiple_files, sender);
+                self.embedder_proxy.send(msg);
+            }
+        }
+    }
```
Comment (paulrouget) [inline components/constellation/constellation.rs]: If we are forwarding a message, maybe we want to keep the same name.

Also, I talked to @cbrewster and we agree that maybe it's better to give access to the EmbedderProxy to each script thread.

File: ports/servo/browser.rs [hunk 6]
```diff
@@ -288,13 +291,49 @@ impl Browser {
...
+            let _ = sender.send(file.map(|x| vec![x]));
+        }
+    }).expect("Thread spawning failed");
+}
+
+#[cfg(not(any(target_os = "macos", target_os = "linux", target_os = "windows")))]
+fn platform_get_selected_files(_patterns: Vec<FilterPattern>,
+                               _multiple_files: bool,
+                               sender: IpcSender<Option<Vec<String>>>) {
+    sender.send(None);
```
Comment (paulrouget) [inline ports/servo/browser.rs]: Maybe `warn!("File picker not implemented")` first?

**Review by paulrouget — CHANGES_REQUESTED** [review]
So I'd really like to see how that would work if we were to share the embedder proxy with the script thread. Sorry, changed my mind again :|

**Comment by kwonoj** [comment]
I'll try to update short-curcuit as suggested. Due to personal schedules I may need some time to wrap things up, once done will ping via @-mention here.

File: components/constellation/constellation.rs [hunk 7]
```diff
@@ -546,7 +546,8 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
           STF: ScriptThreadFactory<Message=Message>
 {
     /// Create a new constellation thread.
-    pub fn start(state: InitialConstellationState) -> (Sender<FromCompositorMsg>, IpcSender<SWManagerMsg>) {
+    pub fn start(state: InitialConstellationState)
+        -> (Sender<FromCompositorMsg>, IpcSender<SWManagerMsg>) {
```
Comment (paulrouget) [inline components/constellation/constellation.rs]: Unnecessary change.

File: components/net/filemanager_thread.rs [hunk 8]
```diff
@@ -159,22 +100,20 @@ impl FileManager {
...
-                    store.select_file(filter, sender, origin, opt_test_path, ui);
+                    store.select_file(filter, sender, origin, opt_test_path, embedder);
                 }).expect("Thread spawning failed");
             }
             FileManagerThreadMsg::SelectFiles(filter, sender, origin, opt_test_paths) => {
                 let store = self.store.clone();
+                let embedder = self.embedder_proxy.clone();
                 thread::Builder::new().name("select files".to_owned()).spawn(move || {
-                    store.select_files(filter, sender, origin, opt_test_paths, ui);
+                    store.select_files(filter, sender, origin, opt_test_paths, embedder );
```
Comment (paulrouget) [inline components/net/filemanager_thread.rs]: Neat: Trailing space.

File: components/net/filemanager_thread.rs [hunk 9]
```diff
@@ -279,21 +218,34 @@ impl FileManagerStore {
...
-                       ui: &UI)
-        where UI: UIProvider,
-    {
+    fn get_selected_files(&self,
+                            patterns: Vec<FilterPattern>,
+                            multiple_files: bool,
+                            embedder_proxy: EmbedderProxy) -> Option<Vec<String>> {
+        if opts::get().headless {
+            return None;
+        }
```
Comment (paulrouget) [inline components/net/filemanager_thread.rs]: Can you let the embedder take the decision to return None when we run in headless mode?

File: components/net/filemanager_thread.rs [hunk 10]
```diff
@@ -279,21 +218,34 @@ impl FileManagerStore {
...
 
-    fn select_file<UI>(&self,
-                       patterns: Vec<FilterPattern>,
-                       sender: IpcSender<FileManagerResult<SelectedFile>>,
-                       origin: FileOrigin,
-                       opt_test_path: Option<String>,
-                       ui: &UI)
-        where UI: UIProvider,
-    {
+    fn get_selected_files(&self,
```
Comment (paulrouget) [inline components/net/filemanager_thread.rs]: Neat: Maybe rename `get_selected_files` to `query_files_from_embedder`.

File: components/net/Cargo.toml [hunk 11]
```diff
@@ -47,9 +48,7 @@ url = "1.2"
 uuid = {version = "0.6", features = ["v4"]}
 webrender_api = {git = "https://github.com/servo/webrender", features = ["ipc"]}
 
-[target.'cfg(any(target_os = "macos", target_os = "linux", target_os = "windows"))'.dependencies]
-tinyfiledialogs = "3.0"
-
 [[test]]
 name = "main"
 path = "tests/main.rs"
+
```
Comment (paulrouget) [inline components/net/Cargo.toml]: Neat: extra line.

File: components/net/tests/fetch.rs [hunk 12]
```diff
@@ -118,7 +118,8 @@ fn test_fetch_blob() {
     use ipc_channel::ipc;
     use net_traits::blob_url_store::BlobBuf;
 
-    let context = new_fetch_context(None);
+    //let (sender, _receiver) = ipc::channel().unwrap();
```
Comment (paulrouget) [inline components/net/tests/fetch.rs]: Remove this comment.

File: components/net/tests/filemanager_thread.rs [hunk 13]
```diff
@@ -2,31 +2,20 @@
...
-    fn open_file_dialog_multi(&self, _path: &str, _patterns: Vec<FilterPattern>) -> Option<Vec<String>> {
-        Some(vec!["tests/test.jpeg".to_string()])
-    }
-}
-
 #[test]
 fn test_filemanager() {
-    let filemanager = FileManager::new();
+    let filemanager = FileManager::new(create_embedder_proxy());
+    PREFS.set("dom.testing.htmlinputelement.select_files.enabled", PrefValue::Boolean(true));
```
Comment (paulrouget) [inline components/net/tests/filemanager_thread.rs]: I'm wondering why this was not necessary before?

File: components/script_traits/script_msg.rs [hunk 14]
```diff
@@ -209,3 +210,10 @@ pub enum SWManagerMsg {
     /// Provide the constellation with a means of communicating with the Service Worker Manager
     OwnSender(IpcSender<ServiceWorkerMsg>),
 }
+
+/// Messages outgoing from the File Manager thread to constellation
+#[derive(Deserialize, Serialize)]
+pub enum FileManagerMsg {
+    /// Requesting to open file select dialog
+    OpenFileSelectDialog(Vec<FilterPattern>, bool, IpcSender<Option<Vec<String>>>)
+}
```
Comment (paulrouget) [inline components/script_traits/script_msg.rs]: I might have missed something, but is that message actually used somewhere?

File: components/net/filemanager_thread.rs [hunk 15]
```diff
@@ -279,21 +218,34 @@ impl FileManagerStore {
...
+                            embedder_proxy: EmbedderProxy) -> Option<Vec<String>> {
+        if opts::get().headless {
+            return None;
+        }
+
+        let (ipc_sender, ipc_receiver) = ipc::channel().expect("Failed to create IPC channel!");
+        let msg = EmbedderMsg::SelectFiles(patterns, multiple_files, ipc_sender);
+
+        embedder_proxy.send(msg);
+        ipc_receiver.recv().unwrap()
```
Comment (paulrouget) [inline components/net/filemanager_thread.rs]: Catch the error here:

```
match ipc_receiver.recv() {
    Err(e) => warn!("Failed to receive files from emebdder ({}).", e),
    ...
}
```

**Review by paulrouget — CHANGES_REQUESTED** [review]
This feels a lot better! I think we made the right decision to use the embedder proxy.

A few neats and then I think we good to go.

Thank you.

File: components/net/tests/filemanager_thread.rs (hunk 13 above)
Comment (kwonoj) [reply components/net/tests/filemanager_thread.rs]: Previously there was mock UI provider in filemanager thread does inject test files instead (https://github.com/servo/servo/pull/20480/files#diff-5a65a0caa6da8aa6e23da0a4101c6a86L19) regardless of testing prefs. Now changed code path removes UI provider from thread and always asks to embedder message, so setting up test to mock fixture for file selection.

File: components/script_traits/script_msg.rs (hunk 14 above)
Comment (kwonoj) [reply components/script_traits/script_msg.rs]: No, I guess it's my mistake leftover while refactoring. Thanks for catching.

File: components/net/filemanager_thread.rs (hunk 9 above)
Comment (kwonoj) [reply components/net/filemanager_thread.rs]: Guess so, will try for it.

**Comment by paulrouget** [comment]
```
error: unused import: `servo_config::opts`
  --> components/net/filemanager_thread.rs:12:5
   |
12 | use servo_config::opts;
   |     ^^^^^^^^^^^^^^^^^^
   |
   = note: `-D unused-imports` implied by `-D warnings`
error: aborting due to previous error
error: Could not compile `net`.
```

**Comment by kwonoj** [comment]
@paulrouget sorry, updated. I think I did update imports, but seems missing somehow.

**Review by paulrouget — CHANGES_REQUESTED** [review]
@bors-servo delegate+

Just a tiny change. Squash and then we good to go.

**Comment by kwonoj** [comment]
@paulrouget Thanks, PR's updated. Seems delegate doesn't triggered bot for some reason?

**Review by paulrouget — APPROVED** [review]

**Comment by kwonoj** [comment]
😓 I can see my fault. Updating now.

**Comment by kwonoj** [comment]
@paulrouget sorry for missing changes - it was leaving config while taking out module imports, now PR updated to reflect it correctly (https://github.com/servo/servo/pull/20480/files#diff-1cdf7ea8e64aef18a475db15965146e0L10)

**Comment by kwonoj** [comment]
Latest change I made was not considering execution path in `browser.rs` makes unused imports on android (specifically, non win/linux/macos target). Updated PR. I tried to verify build locally as much but didn't have android build setup on my machines 😭

**Comment by kwonoj** [comment]
PR updated again to resolve another android build failure.

**Comment by nox** [comment]
@bors-servo r-

Sorry @paulrouget but the CI is on fire and #20693 needs to land first.

**Comment by nox** [comment]
Homu is super confused right now, will reopen the PR once the other one finally lands.

**Comment by paulrouget** [comment]
@bors-servo retry

- Infra: Servers failed to start.
